## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.nunique()


first            360
last             477
job              498
phone            999
email            999
client_id        999
order_id       10000
order_date       150
order_week        23
order_year         2
item_id          500
category           5
subcategory       18
unit_price       490
unit_cost        487
unit_weight      393
qty             2911
line_number       10
dtype: int64

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.isna()


first   last    job  phone  email  client_id  order_id  order_date  \
0      False  False  False  False  False      False     False       False   
1      False  False  False  False  False      False     False       False   
2      False  False  False  False  False      False     False       False   
3      False  False  False  False  False      False     False       False   
4      False  False  False  False  False      False     False       False   
...      ...    ...    ...    ...    ...        ...       ...         ...   
54634  False  False  False  False  False      False     False       False   
54635  False  False  False  False  False      False     False       False   
54636  False  False  False  False  False      False     False       False   
54637  False  False  False  False  False      False     False       False   
54638  False  False  False  False  False      False     False       False   

       order_week  order_year  item_id  category  subcategory  unit_price  \
0           False       False    False     False        False       False   
1           False       False    False     False        False       False   
2           False       False    False     False        False       False   
3           False       False    False     False        False       False   
4           False       False    False     False        False       False   
...           ...         ...      ...       ...          ...         ...   
54634       False       False    False     False        False       False   
54635       False       False    False     False        False       False   
54636       False       False    False     False        False       False   
54637       False       False    False     False        False       False   
54638       False       False    False     False        False       False   

       unit_cost  unit_weight    qty  line_number  
0          False        False  False        False  
1          False        False  False        False  
2          False        False  False        False  
3          False        False  False        False  
4          False        False  False        False  
...          ...          ...    ...          ...  
54634      False        False  False        False  
54635      False        False  False        False  
54636      False        False  False        False  
54637      False        False  False        False  
54638      False        False  False        False  

[54639 rows x 18 columns]

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.isna().count()


first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [7]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.dtypes


first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [8]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.min()


first                             Aaron
last                             Abbott
job               Accommodation manager
phone                     (203)310-1717
email          a.garcia.3072@durham.com
client_id                         10033
order_id                        1000886
order_date                   2023-01-01
order_week                            1
order_year                         2022
item_id                  AAD15703-65-3A
category                    consumables
subcategory                  accounting
unit_price                         0.01
unit_cost                          0.01
unit_weight                         0.0
qty                                   0
line_number                           0
dtype: object

In [9]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.max()


first                              Zachary
last                                Zavala
job                           Youth worker
phone                    999.801.2159x2255
email          zacharyjohnson@jacobson.net
client_id                            99984
order_id                           9998480
order_date                      2023-05-30
order_week                              52
order_year                            2023
item_id                     ZZU43604-44-9M
category                          software
subcategory                       wall art
unit_price                         1396.23
unit_cost                           846.27
unit_weight                          46.43
qty                                3958244
line_number                              9
dtype: object

In [10]:
# What three item categories had the most entries? #value_count automatically sorts
df.category.value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [11]:
# For the category with the most entries,
# which subcategory had the most entries?
df.subcategory.value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [12]:
# Which five clients had the most entries in the data?

df.client_id.value_counts().head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [13]:
# Store the client ids of those top 5 clients in a list. 

top_5_client_list = df.client_id.value_counts().head().index.tolist()
top_5_client_list

[33615, 66037, 46820, 38378, 24741]

In [14]:
# How many total units (the qty column) did the
# client with the most entries order order?
# This will be a combination of dataframe indexing approaches i.e. applying the loc indexer
# and using the index attribute for column attribute i.e. df.loc[df.index == value, 'qty'].sum()

df.loc[df.client_id == 33615, 'qty'].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [15]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df["line_subtotal"] = df.unit_price * df.qty
df[["unit_price","qty","line_subtotal"]].head(2)


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [16]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.


order_weight_price = df.apply(lambda order: order['unit_weight']*7 
                              if order['unit_weight']* order ['qty']>50 
                              else order['unit_weight']*10, axis=1)
df['shipping_price'] = order_weight_price * df ['qty']
df[['unit_price','unit_weight','qty','shipping_price']].head(2)




unit_price  unit_weight  qty  shipping_price
0     1096.80         7.50  105          5512.5
1       24.95         1.49   21           312.9

In [17]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
#PS: used line_price instead of total_price (since output example showed line_price and the following steps ask for line_price not total_price from hereon


df['line_price'] = (df['line_subtotal'] + df['shipping_price']) * 1.0925
df['line_price'] = df['line_price'].round(2)
df[['line_subtotal', 'shipping_price','line_price']].head(3)




line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [18]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_cost'] = df['unit_cost']* df['qty'] + df['shipping_price']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   

   unit_weight  qty  line_number  line_subtotal  shipping_price  line_price  \
0         7.50  105            1      115164.00         5512.50   131839.08   
1         1.49   21            0         523.95          312.90      914.26   
2         1.68   39            6         527.28          458.64     1077.12   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 22 columns]

In [19]:
# Create a column for the profit of
# each line using line cost and line price


df['line_profit'] = df['line_price'] - df['line_cost']
df ['line_profit'] = df['line_profit'].round(2)
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   

   line_number  line_subtotal  shipping_price  line_price  line_cost  \
0            1      115164.00         5512.50   131839.08   85597.05   
1            0         523.95          312.90      914.26     629.79   
2            6         527.28          458.64     1077.12     765.18   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  

[3 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [20]:
# Check your work using the totals above

order_ids = [2742071,2173913,6128929]

def find_total_price(order_id):
    total_price_df = df.loc[df["order_id"] == order_id,'line_price']
    return round(total_price_df.sum(),2)

for order_id in order_ids:
    print (f"Order ID {order_id} Total Price ${find_total_price(order_id)}")



Order ID 2742071 Total Price $152811.89
Order ID 2173913 Total Price $162388.71
Order ID 6128929 Total Price $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [21]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.


def client_totals(client_id,column):    
    client_df = df.loc[df.client_id == client_id, column]
    total = round(client_df.sum(),2)   
    return total     

for client_id in top_5_client_list:     #top_5_client_list pulled from top part 1 variable assignment
    print(f"{client_id} ${client_totals(client_id,'line_price'):.2f}")




33615 $8377308.52
66037 $10259514.79
46820 $9743794.36
38378 $12906550.87
24741 $82268892.02


In [22]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



top_5_client_id = {'client_id': top_5_client_list}
                
columns = ['qty', 'shipping_price', 'line_price', 'line_cost', 'line_profit']

for column in columns:
    
    top_5_client_id[column] = [client_totals(client_id,column) if column != 'qty' else int(client_totals(client_id,column)) for client_id in top_5_client_list ]
    
summarized_df = pd.DataFrame(top_5_client_id).sort_values('line_profit', ascending=False)

summarized_df




client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [23]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

format_columns = ['shipping_price', 'line_price', 'line_cost', 'line_profit']

def format(amount):
    return f"${amount/1000000:.2f}M"                          

for col in format_columns:
    summarized_df[col] = summarized_df[col].apply(format)
    
summarized_df.head()


rename = {'client_id':'Client ID','qty':'Units','shipping_price':'Shipping','line_price':'Total Revenue','line_cost':'Total Cost','line_profit':'Total Profit',}
                                            
summarized_df_rename = summarized_df.rename(columns=rename)

summarized_df_rename.head()


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M

In [24]:
# Sort the updated data by "Total Profit" form highest to lowest

summarized_df_sorted = summarized_df_rename.sort_values(by='Total Profit', ascending=False)
summarized_df_sorted.head()


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M